src/App.tsx

```diff
@@ -1,24 +1,184 @@
+import { ErrorNotification } from './components/ErrorNotification';
 
-const USER_ID = 0;
+const USER_ID = 12042;
 
 export const App: React.FC = () => {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [error, setError] = useState<string | null>(null);
+  const [status, setStatus] = useState<Status>(Status.all);
+  const [todoTitle, setTodoTitle] = useState<string>('');
+  const [loadingTodoIds, setLoadingTodoIds] = useState<number[]>([]);
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+
+  const titleField = useRef<HTMLInputElement>(null);
+
+  const handleError = (errorMessage: string) => {
+    setError(errorMessage);
+    setTimeout(() => {
+      setError(null);
+    }, 3000);
+  };
+
+  useEffect(() => {
+    if (titleField.current) {
+      titleField.current.focus();
+    }
+
+    todosService.getTodos(USER_ID)
+      .then(setTodos)
+      .catch(() => (handleError('Unable to load todos')));
+  }, []);
+
+  const filteredTodos = useMemo(() => {
+    switch (status) {
+      case Status.all: {
+        return todos;
+      }
+
+      case Status.active: {
+        return todos.filter(todo => !todo.completed);
+      }
+
+      case Status.completed: {
+        return todos.filter(todo => todo.completed);
+      }
+
+      default: {
+        return todos;
+      }
+    }
+  }, [todos, status]);
+
+  const createTodo = (event: React.FormEvent<HTMLFormElement>) => {
+    event.preventDefault();
+    if (!todoTitle) {
+      handleError('Title should not be empty');
+    }
+
+    if (todoTitle) {
+      const tempTodoId = 0;
+
+      setTempTodo({
+        id: tempTodoId,
+        userId: USER_ID,
+        title: todoTitle,
+        completed: false,
+      });
+
+      setLoadingTodoIds(curr => [...curr, tempTodoId]);
+
+      todosService.addTodo({
+        userId: USER_ID, title: todoTitle, completed: false,
+      })
+        .then(newTodo => {
+          setTodos(currentTodos => {
+            const maxId = Math.max(0, ...currentTodos.map(todo => todo.id));
+            const id = maxId + 1;
+
+            return [...currentTodos, { ...newTodo, id }];
+          });
+          setTodoTitle('');
+        })
+        .catch(() => handleError('Unable to add a todo'))
+        .finally(() => {
+          setTempTodo(null);
+          setLoadingTodoIds(curr => curr.filter(id => id !== tempTodoId));
+        });
+    }
+  };
+
+  const deleteTodo = (todoId: number) => {
+    setLoadingTodoIds(curr => [...curr, todoId]);
+
+    todosService.deleteTodo(todoId)
+      .then(() => setTodos(
+        currentTodos => currentTodos.filter(todo => todo.id !== todoId),
+      ))
+      .catch(() => handleError('Unable to delete a todo'))
+      .finally(() => {
+        setLoadingTodoIds(curr => curr.filter(id => id !== todoId));
+      });
+  };
+
+  const clearTodos = () => {
+    todos.map(todo => {
+      if (todo.completed === true) {
+        todosService.deleteTodo(todo.id);
+      }
+
+      return 1;
```

Comment:
```suggestion
    todos.forEach(todo => {
      if (todo.completed) {
        todosService.deleteTodo(todo.id);
      }
```
- if you don't need a new array you can just iterate it with forEach
- you don't need to compare with true
- and you don't need to return something if you ignore this values